lib/Sema/CSSimplify.cpp

```diff
@@ -7566,10 +7566,16 @@ ConstraintSystem::simplifyKeyPathConstraint(
     }
 
     if (auto fnTy = type->getAs<FunctionType>()) {
-      definitelyFunctionType = true;
+      if (fnTy->getParams().size() != 1) {
+        if (!shouldAttemptFixes())
+          return false;
 
-      if (fnTy->getParams().size() != 1)
-        return false;
+        // Fix function type mismatch with an IgnoreContextualType fix
+        // and proceed as if the constraint got resolved.
+        resolveAsFunctionTypeMismatchFix = true;
+        return true;
```

Comment:
Why not record fix inline here and return `false` if that fails? That way we don't have to drag `resolveAsFunctionTypeMismatchFix` flag though the whole function.

Reply:
I understand that this is because `resolveKPTy` gets formed later but I think it's okay to duplicate that bit here.

Reply:
Yeah, there seems to be quite a bit of code to determine the kind of `kpDecl` to use to create the `resolveKPTy` which I wanted to avoid duplicating. I'll see if I can factor it out instead.

Reply:
Let's do this - just form it as a read-only key path and add a special diagnostic in this case e.g. `cannot convert multi-argument <function type> into a key path`? It seems like what ever this key path gets resolved to the main issue here is with function type. WDYT?